src/format/Quoting.cc

```diff
@@ -39,6 +40,14 @@ Format::QuoteUrlEncodeUsername(const char *name)
     return QuoteMimeBlob(name);
 }
 
+#if USE_IDENT
+char *
+Format::QuoteUrlEncodeUsername(const Ident::User &ident)
+{
+    return ident ? QuoteUrlEncodeUsername(SBuf(ident.value()).c_str()) : nullptr;
```

Comment:
I think the `SBuf(...).c_str()` construction should be safe in regard to SBuf() lifetime:

 [Temporary objects](https://eel.is/c++draft/class.temporary#4) are destroyed as the last step in evaluating the full-expression. Expressions are defined [here](https://eel.is/c++draft/intro.execution#5.6), for example:
```
  if (S(3).v())                 // full-expression includes lvalue-to-rvalue and int to bool conversions,
                                // performed before temporary is deleted at end of full-expression
```
Here is the rule for `return` statement:
[(6.11)](https://eel.is/c++draft/class.temporary#6.11)
The lifetime of a temporary bound to the returned value in a function return statement ([[stmt.return]](https://eel.is/c++draft/stmt.return)) is not extended; the temporary is destroyed at the end of the full-expression in the return statement[.](https://eel.is/c++draft/class.temporary#6.11.sentence-1).

So in our case the SBuf() temporary should be destroyed after the entire return value is constructed (i.e.,  after QuoteUrlEncodeUsername() finishes).

One should be mindful of casts [applied](https://stackoverflow.com/questions/28644398/ternary-operator-and-prolonging-the-lifetime-of-a-temporary-object-via-reference) to temporary variables.

Reply:
I agree that the destruction rules for the full-expression allow us to do what PR code is doing. Thank you for finding them.